docs/advanced.rst

```diff
@@ -243,3 +241,170 @@ the functions defined in their class body.
+Once we define a message part, we need to tell the engine how to use it. Since message parts are designed to provide
+engine-specific metadata while maintaining cross-engine compatibility, engines should check for the classes they know
+and cast unknown classes to a string. This lets parts control how they will be serialized in engines that don't natively
+support them.
+
+A common, but not necessary, pattern is to write an engine that *wraps* other engines. This wrapper engine acts as a
+translation layer for specific message parts, eventually compiling a single string (or list of strings) for a base
+engine (e.g. LlamaEngine) to consume.
+
+Here's the implementation for our chain-of-thought example. Note how we translate the parts by building a new message
+with a string content composed of parts in ``translate_message``, then use that method in our implementations of
+``message_len`` and ``predict``.
+
+.. code-block:: python
+
+    from kani import AIFunction, ChatMessage, MessagePart
+    from kani.engines.base import BaseEngine, Completion
+
+    class ChainOfThoughtEngine(BaseEngine):
+        def __init__(self, engine: BaseEngine):
+            self.engine = engine
+            self.max_context_size = engine.max_context_size
+
+        @staticmethod
+        def translate_message(message: ChatMessage) -> ChatMessage:
+            """Translate a input message into a simple string-only message to pass to the underlying engine."""
+            content = ""
+            # iterate over parts: if it's a thought, place its data in the content; otherwise prefix it with "Answer: "
+            for part in message.parts:
+                if isinstance(part, ThoughtPart):
+                    content += f"{part.data}\n"
+                else:
+                    content += f"Answer: {str(part)}"
+            # return the translated message
+            return message.copy_with(content=content.strip())
+
+        # === BaseEngine interface ===
+        def message_len(self, message: ChatMessage) -> int:
+            return self.engine.message_len(self.translate_message(message))
+
+        async def predict(
+            self, messages: list[ChatMessage], functions: list[AIFunction] | None = None, **hyperparams
+        ) -> Completion:
+            # translate the messages
+            translated_messages = [self.translate_message(m) for m in messages]
+
+            # generate a completion using the underlying engine given those messages
+            result = await self.engine.predict(translated_messages, functions, **hyperparams)
+
+            # parse the string-completion back into parts - with some checks to make sure the model actually output the
+            # right format
+            text = result.message.text
+            if "Answer:" in text:
+                thought, answer = text.split("Answer:", 1)
+                message_with_parts = result.message.copy_with(parts=[ThoughtPart(data=thought.strip()), answer.strip()])
```

Comment:
Does this put all thoughts into one part? (Is that expected behavior?) or should we also split on newline